File: packages/ai-gateway/src/adapters/openai-compatible.ts
// Cliente OpenAI-compatible: base compartida por OpenAI y OpenRouter.
// Convierte el contrato interno (ChatRequest) al formato /v1/chat/completions.

import { postJson, ProviderHttpError } from "./http.js";
import type { ChatRequest, ChatResult, ModelProvider, ProviderId } from "../types.js";

export interface OpenAiCompatibleOptions {
  baseUrl: string;
  extraHeaders?: Record<string, string>;
  label: string;
}

export class OpenAiCompatibleProvider implements ModelProvider {
  readonly id: ProviderId;

  constructor(
    id: ProviderId,
    private readonly models: string[],
    private readonly opts: OpenAiCompatibleOptions,
    private readonly resolveKey: (req: ChatRequest) => string | undefined,
  ) {
    this.id = id;
  }

  listModels(): string[] {
    return [...this.models];
  }

  async chat(req: ChatRequest): Promise<ChatResult> {
    const apiKey = this.resolveKey(req);
    if (!apiKey) throw new Error(`Sin API key para ${this.opts.label} (setea la env o usa BYOK)`);

    const messages = req.messages.map((m) => ({ role: m.role, content: m.content }));

    const data = await postJson<{
      choices?: Array<{ message?: { content?: string } }>;
      error?: { message?: string };
      usage?: { prompt_tokens?: number; completion_tokens?: number };
    }>({
      url: `${this.opts.baseUrl}/chat/completions`,
      body: {
        model: req.model,
        messages,
        stream: false,
        temperature: req.temperature ?? 0.7,
        ...(req.maxTokens ? { max_tokens: req.maxTokens } : {}),
      },
      headers: {
        authorization: `Bearer ${apiKey}`,
        "x-provider": this.opts.label,
        ...this.opts.extraHeaders,
      },
      signal: req.signal,
    });

    const content = data.choices?.[0]?.message?.content;
    if (content === undefined || content === null) {
      throw new ProviderHttpError(`Respuesta vacía de ${this.opts.label}`, 0, this.opts.label);
    }

    return {
      content,
      model: req.model,
      provider: this.id,
      usage: {
        inputTokens: data.usage?.prompt_tokens ?? 0,
        outputTokens: data.usage?.completion_tokens ?? 0,
      },
    };
  }
}

/** Key del usuario (BYOK) o de la plataforma según env var. */
export function keyFromEnv(envName: string) {
  return (req: ChatRequest): string | undefined => {
    if (req.apiKey) return req.apiKey;
    return process.env[envName];
  };
}

File: apps/api/internal/gateway/adapters/openai_compatible.go
package adapters

import (
	"fmt"
	"os"

	"github.com/blu-ia/api/internal/gateway"
)

type OpenAICompatibleOptions struct {
	BaseURL      string
	ExtraHeaders map[string]string
	Label        string
}

type OpenAICompatibleProvider struct {
	id         gateway.ProviderID
	models     []string
	opts       OpenAICompatibleOptions
	resolveKey func(req gateway.ChatRequest) *string
}

func NewOpenAICompatible(id gateway.ProviderID, models []string, opts OpenAICompatibleOptions, resolveKey func(gateway.ChatRequest) *string) *OpenAICompatibleProvider {
	return &OpenAICompatibleProvider{id: id, models: models, opts: opts, resolveKey: resolveKey}
}

func (p *OpenAICompatibleProvider) ID() gateway.ProviderID { return p.id }
func (p *OpenAICompatibleProvider) ListModels() []string   { return append([]string(nil), p.models...) }

func (p *OpenAICompatibleProvider) Chat(req gateway.ChatRequest) (gateway.ChatResult, error) {
	apiKey := p.resolveKey(req)
	if apiKey == nil || *apiKey == "" {
		return gateway.ChatResult{}, fmt.Errorf("Sin API key para %s (setea la env o usa BYOK)", p.opts.Label)
	}
	messages := make([]map[string]string, 0, len(req.Messages))
	for _, m := range req.Messages {
		messages = append(messages, map[string]string{"role": m.Role, "content": m.Content})
	}
	body := map[string]any{
		"model":    req.Model,
		"messages": messages,
		"stream":   false,
	}
	if req.Temperature != nil {
		body["temperature"] = *req.Temperature
	} else {
		body["temperature"] = 0.7
	}
	if req.MaxTokens != nil {
		body["max_tokens"] = *req.MaxTokens
	}
	headers := map[string]string{
		"authorization": "Bearer " + *apiKey,
		"x-provider":    p.opts.Label,
	}
	for k, v := range p.opts.ExtraHeaders {
		headers[k] = v
	}
	type respBody struct {
		Choices []struct {
			Message *struct {
				Content *string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
		Usage *struct {
			PromptTokens     *int `json:"prompt_tokens"`
			CompletionTokens *int `json:"completion_tokens"`
		} `json:"usage"`
		Error *struct {
			Message string `json:"message"`
		} `json:"error"`
	}
	data, err := PostJSON[respBody](PostJsonOptions{
		URL:     p.opts.BaseURL + "/chat/completions",
		Body:    body,
		Headers: headers,
		Ctx:     req.Ctx,
	})
	if err != nil {
		return gateway.ChatResult{}, err
	}
	if len(data.Choices) == 0 || data.Choices[0].Message == nil || data.Choices[0].Message.Content == nil {
		return gateway.ChatResult{}, &ProviderHttpError{Message: fmt.Sprintf("Respuesta vacía de %s", p.opts.Label), Status: 502, ProviderLabel: p.opts.Label}
	}
	content := *data.Choices[0].Message.Content
	var inTok, outTok int
	if data.Usage != nil {
		if data.Usage.PromptTokens != nil {
			inTok = *data.Usage.PromptTokens
		}
		if data.Usage.CompletionTokens != nil {
			outTok = *data.Usage.CompletionTokens
		}
	}
	return gateway.ChatResult{
		Content:  content,
		Model:    req.Model,
		Provider: p.id,
		Usage:    struct{ InputTokens int `json:"inputTokens"`; OutputTokens int `json:"outputTokens"` }{InputTokens: inTok, OutputTokens: outTok},
	}, nil
}

func KeyFromEnv(envName string) func(gateway.ChatRequest) *string {
	return func(req gateway.ChatRequest) *string {
		if req.APIKey != nil && *req.APIKey != "" {
			return req.APIKey
		}
		if v := os.Getenv(envName); v != "" {
			return &v
		}
		return nil
	}
}
